components/tidb_query_expr/src/impl_string.rs

```diff
@@ -304,6 +304,37 @@ pub fn rpad(arg: BytesRef, len: &Int, pad: BytesRef, writer: BytesWriter) -> Res
     }
 }
 
+#[rpn_fn(writer)]
+#[inline]
+pub fn rpad_utf8(
+    arg: BytesRef,
+    len: &Int,
+    pad: BytesRef,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    let input = match str::from_utf8(&*arg) {
+        Ok(arg) => arg,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let pad = match str::from_utf8(&*pad) {
+        Ok(pad) => pad,
+        Err(err) => return Err(box_err!("invalid input value: {:?}", err)),
+    };
+    let input_len = input.chars().count();
+    match validate_target_len_for_pad(*len < 0, *len, input_len, 4, pad.is_empty()) {
+        None => Ok(writer.write(None)),
+        Some(0) => Ok(writer.write_ref(Some(b""))),
+        Some(target_len) => {
+            let r = input
+                .chars()
+                .chain(pad.chars().cycle())
+                .take(target_len)
+                .collect::<String>();
+            Ok(writer.write(Some(r.into_bytes())))
```

Comment:
We can use `write_ref` instead.

Reply:
I tried using `write_ref` but it didn't work because we create a new (owned) string with `chain` and `take` and `collect`. Not sure if I missed something though.

Reply:
Can we reduce the allocation here? For exmaple, we can calculate the length first and call `write_partial` to write them without extra allocation.

Reply:
Ah okay got it :)
Should I change the implementation of `rpad()` to use `partial_write()` as well?

Reply:
Yes, I suggest doing so.

Reply:
Hi, I removed the allocations for both `rpad` and `lpad` functions with the latest commit. The resulting code is not as nice as it was with iterators. Any ideas on how to improve this or should we leave it as it is?

Reply:
I think we could leave it as-is.